src/Generator.cpp

```diff
@@ -21,362 +26,655 @@
-        event.fillHit(id_cell.first, enoise);
-        energyrec += enoise;
-        if (debug) std::cout << "adding " << enoise << " GeV noise in cell " << std::endl;      
-      }
+    std::unordered_map<uint32_t, TH1F> hCellEnergyMap;
+    std::unordered_map<uint32_t, TH1F> hCellEnergyEvtMap;
+
+    TStopwatch t;
+    t.Start();
+
+    // some initializations
+    double energygen=0.;
+    double energygenincells=0.;
+    double energyrec=0.;
+
+    // Geometry
+    cout<<"I'm building the geometry, please wait..." <<endl;
+
+    // Shower parametrization
+    ShowerParametrization aShowerParametrization(parameters_.shower());
+
+    // Canvas for the geometry map in rootfile
+    std::vector<std::unique_ptr<TCanvas>> canvas;
+
+
+    // Build geometry
+    int layer_min, layer_max;
+    int display_layer;
+
+    if (parameters_.geometry().layer==-1) {
+        layer_min = 0;
+        layer_max = parameters_.geometry().layers_z.size();
+        display_layer = parameters_.display().layer;
     }
-      
-    // generate energy spots according to transverse profile
-    double r0 = aShowerParametrization.r0(parameters_.geometry().layer);
-    // take longitudinal profile as mean energy per layer for fixed energy
-    double layer_weight = 1.;
-    if (parameters_.geometry().layer!=-1) layer_weight = aShowerParametrization.getLayerProfile()[parameters_.geometry().layer];
-
-    // energy spot 
-    // no fluctuations: fixed energy = 1. / nhitspergev
-    // fluctuations: alpha/sqrt(E) -> Poissonian nbr hits of energy 1/alpha^2
-    // where alpha is the stochastic term of the resolution
-    int nhits;
-    double denrj;
-    if (!parameters_.generation().fluctuation) {
-      nhits = int(parameters_.generation().energy*layer_weight*parameters_.generation().number_of_hits_per_gev);
-      denrj = 1./parameters_.generation().number_of_hits_per_gev;
-    } else {
-      denrj = aShowerParametrization.spotEnergy();
-      nhits = gun_.Poisson(parameters_.generation().energy*layer_weight/denrj); 
-    }  
-
-    if (debug) {
-      std::cout << " number of generated hits " << nhits << " with energy " << denrj <<
-        std::endl;
+    else {
+        layer_min = parameters_.geometry().layer;
+        layer_max = parameters_.geometry().layer + 1;
+        display_layer = layer_min;
     }
 
+    std::vector<Cell> cell_collection[52];
+
+    for (int layer_id = layer_min; layer_id < layer_max; layer_id++) {
+
+        cout << "=====> Layer " <<layer_id + 1 << " / "<< layer_max << '\r' << flush;
+
+        if (parameters_.geometry().type!=Parameters::Geometry::Type::External)
+            geometry_.constructFromParameters(parameters_.general().debug, layer_id, display_layer);
+        else
+            geometry_.constructFromJson(parameters_.general().debug, layer_id);
+
+        for (auto& cell : geometry_.getCells()){
+            if (cell.second.getLayer() == layer_id)
+                cell_collection[layer_id].push_back(cell.second);
```

Comment:
Why copying all the cells locally, since they are already all stored and accessible in the geometry class (which is there for this reason)?
The best is probably to implement inside the geometry class the possibility to retrieve cells from a given layer.

Reply:
The copy is mandatory since the all layers implementation because the geometry is built layer after layer. For example if the user wants build layer 1 to 3, Generator builds layer 1 first. When it built layer 2 the cells collection was reset because it is a new call to geometry. Same situation when you build the layer 3. So if you don't copy the cells collection after each layer building your cells collection after the entire geometry building corresponds to the last layer builds (layer 3 in the example).

Reply:
In that case, it would be better to change the geometry class, and have one single geometry object storing all the cells and with the proper interface to be able to retrieve cells in a given layer.